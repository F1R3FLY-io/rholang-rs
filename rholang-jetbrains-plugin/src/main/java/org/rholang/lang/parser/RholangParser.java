package org.rholang.lang.parser;

import com.intellij.lang.ASTNode;
import com.intellij.lang.PsiBuilder;
import com.intellij.lang.PsiParser;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;

public class RholangParser implements PsiParser {
    private static final Logger LOG = Logger.getInstance(RholangParser.class);

    @NotNull
    @Override
    public ASTNode parse(@NotNull IElementType root, @NotNull PsiBuilder builder) {
        PsiBuilder.Marker rootMarker = builder.mark();

        // Get the full text from the builder
        String text = builder.getOriginalText().toString();

        // Check if the code is valid using the Rust parser
        boolean isValid = RholangParserCli.isValid(text);
        LOG.debug("Rholang code is valid: " + isValid);

        // For now, just consume all tokens
        // In a more complete implementation, we would parse the tree returned by the Rust parser
        // and build a corresponding PSI tree
        while (!builder.eof()) {
            builder.advanceLexer();
        }

        rootMarker.done(root);
        return builder.getTreeBuilt();
    }
}
