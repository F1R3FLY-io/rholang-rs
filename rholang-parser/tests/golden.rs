use rholang_parser::RholangParser;
use rstest::rstest;
use std::fs;
use std::path::PathBuf;

#[rstest]
fn golden_test(
    #[base_dir = "tests/corpus/"]
    #[files("*.rho")]
    path: PathBuf,
) {
    let mut settings = insta::Settings::new();
    settings.set_snapshot_path("corpus/golden_snapshots");

    settings.bind(|| {
        let name = path.file_stem().unwrap().to_string_lossy().to_string();

        let parser = RholangParser::new();
        let input = fs::read_to_string(path).expect("Failed to read input file");
        let result = parser.parse(input.as_str());

        // Each snapshot is named after the input file
        insta::assert_debug_snapshot!(name, result);
    })
}
