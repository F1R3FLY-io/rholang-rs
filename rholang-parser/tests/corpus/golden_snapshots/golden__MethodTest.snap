---
source: rholang-parser/tests/golden.rs
expression: result
---
Good(
    [
        AnnProc {
            proc: Contract {
                name: NameVar(
                    Id(
                        Id {
                            name: "MethodTest",
                            pos: SourcePos {
                                line: 3,
                                col: 10,
                            },
                        },
                    ),
                ),
                formals: Names {
                    names: [],
                    remainder: None,
                },
                body: AnnProc {
                    proc: Par {
                        left: AnnProc {
                            proc: Par {
                                left: AnnProc {
                                    proc: Par {
                                        left: AnnProc {
                                            proc: Par {
                                                left: AnnProc {
                                                    proc: Method {
                                                        receiver: AnnProc {
                                                            proc: LongLiteral(
                                                                5,
                                                            ),
                                                            span: SourceSpan {
                                                                start: SourcePos {
                                                                    line: 4,
                                                                    col: 7,
                                                                },
                                                                end: SourcePos {
                                                                    line: 4,
                                                                    col: 8,
                                                                },
                                                            },
                                                        },
                                                        name: Id {
                                                            name: "add",
                                                            pos: SourcePos {
                                                                line: 4,
                                                                col: 9,
                                                            },
                                                        },
                                                        args: [
                                                            AnnProc {
                                                                proc: LongLiteral(
                                                                    5,
                                                                ),
                                                                span: SourceSpan {
                                                                    start: SourcePos {
                                                                        line: 4,
                                                                        col: 13,
                                                                    },
                                                                    end: SourcePos {
                                                                        line: 4,
                                                                        col: 14,
                                                                    },
                                                                },
                                                            },
                                                            AnnProc {
                                                                proc: LongLiteral(
                                                                    5,
                                                                ),
                                                                span: SourceSpan {
                                                                    start: SourcePos {
                                                                        line: 4,
                                                                        col: 15,
                                                                    },
                                                                    end: SourcePos {
                                                                        line: 4,
                                                                        col: 16,
                                                                    },
                                                                },
                                                            },
                                                        ],
                                                    },
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 4,
                                                            col: 7,
                                                        },
                                                        end: SourcePos {
                                                            line: 4,
                                                            col: 17,
                                                        },
                                                    },
                                                },
                                                right: AnnProc {
                                                    proc: Method {
                                                        receiver: AnnProc {
                                                            proc: LongLiteral(
                                                                5,
                                                            ),
                                                            span: SourceSpan {
                                                                start: SourcePos {
                                                                    line: 6,
                                                                    col: 7,
                                                                },
                                                                end: SourcePos {
                                                                    line: 6,
                                                                    col: 8,
                                                                },
                                                            },
                                                        },
                                                        name: Id {
                                                            name: "subtract",
                                                            pos: SourcePos {
                                                                line: 6,
                                                                col: 9,
                                                            },
                                                        },
                                                        args: [
                                                            AnnProc {
                                                                proc: LongLiteral(
                                                                    5,
                                                                ),
                                                                span: SourceSpan {
                                                                    start: SourcePos {
                                                                        line: 6,
                                                                        col: 18,
                                                                    },
                                                                    end: SourcePos {
                                                                        line: 6,
                                                                        col: 19,
                                                                    },
                                                                },
                                                            },
                                                        ],
                                                    },
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 6,
                                                            col: 7,
                                                        },
                                                        end: SourcePos {
                                                            line: 6,
                                                            col: 20,
                                                        },
                                                    },
                                                },
                                            },
                                            span: SourceSpan {
                                                start: SourcePos {
                                                    line: 4,
                                                    col: 5,
                                                },
                                                end: SourcePos {
                                                    line: 6,
                                                    col: 22,
                                                },
                                            },
                                        },
                                        right: AnnProc {
                                            proc: Method {
                                                receiver: AnnProc {
                                                    proc: ProcVar(
                                                        Id(
                                                            Id {
                                                                name: "balances",
                                                                pos: SourcePos {
                                                                    line: 8,
                                                                    col: 7,
                                                                },
                                                            },
                                                        ),
                                                    ),
                                                    span: SourceSpan {
                                                        start: SourcePos {
                                                            line: 8,
                                                            col: 7,
                                                        },
                                                        end: SourcePos {
                                                            line: 8,
                                                            col: 15,
                                                        },
                                                    },
                                                },
                                                name: Id {
                                                    name: "get",
                                                    pos: SourcePos {
                                                        line: 8,
                                                        col: 16,
                                                    },
                                                },
                                                args: [
                                                    AnnProc {
                                                        proc: LongLiteral(
                                                            5,
                                                        ),
                                                        span: SourceSpan {
                                                            start: SourcePos {
                                                                line: 8,
                                                                col: 20,
                                                            },
                                                            end: SourcePos {
                                                                line: 8,
                                                                col: 21,
                                                            },
                                                        },
                                                    },
                                                ],
                                            },
                                            span: SourceSpan {
                                                start: SourcePos {
                                                    line: 8,
                                                    col: 7,
                                                },
                                                end: SourcePos {
                                                    line: 8,
                                                    col: 22,
                                                },
                                            },
                                        },
                                    },
                                    span: SourceSpan {
                                        start: SourcePos {
                                            line: 4,
                                            col: 5,
                                        },
                                        end: SourcePos {
                                            line: 8,
                                            col: 24,
                                        },
                                    },
                                },
                                right: AnnProc {
                                    proc: Method {
                                        receiver: AnnProc {
                                            proc: ProcVar(
                                                Id(
                                                    Id {
                                                        name: "balances",
                                                        pos: SourcePos {
                                                            line: 10,
                                                            col: 7,
                                                        },
                                                    },
                                                ),
                                            ),
                                            span: SourceSpan {
                                                start: SourcePos {
                                                    line: 10,
                                                    col: 7,
                                                },
                                                end: SourcePos {
                                                    line: 10,
                                                    col: 15,
                                                },
                                            },
                                        },
                                        name: Id {
                                            name: "update",
                                            pos: SourcePos {
                                                line: 10,
                                                col: 16,
                                            },
                                        },
                                        args: [
                                            AnnProc {
                                                proc: StringLiteral(
                                                    "random",
                                                ),
                                                span: SourceSpan {
                                                    start: SourcePos {
                                                        line: 10,
                                                        col: 23,
                                                    },
                                                    end: SourcePos {
                                                        line: 10,
                                                        col: 31,
                                                    },
                                                },
                                            },
                                            AnnProc {
                                                proc: LongLiteral(
                                                    5,
                                                ),
                                                span: SourceSpan {
                                                    start: SourcePos {
                                                        line: 10,
                                                        col: 33,
                                                    },
                                                    end: SourcePos {
                                                        line: 10,
                                                        col: 34,
                                                    },
                                                },
                                            },
                                        ],
                                    },
                                    span: SourceSpan {
                                        start: SourcePos {
                                            line: 10,
                                            col: 7,
                                        },
                                        end: SourcePos {
                                            line: 10,
                                            col: 35,
                                        },
                                    },
                                },
                            },
                            span: SourceSpan {
                                start: SourcePos {
                                    line: 4,
                                    col: 5,
                                },
                                end: SourcePos {
                                    line: 10,
                                    col: 37,
                                },
                            },
                        },
                        right: AnnProc {
                            proc: Method {
                                receiver: AnnProc {
                                    proc: ProcVar(
                                        Id(
                                            Id {
                                                name: "balances",
                                                pos: SourcePos {
                                                    line: 12,
                                                    col: 7,
                                                },
                                            },
                                        ),
                                    ),
                                    span: SourceSpan {
                                        start: SourcePos {
                                            line: 12,
                                            col: 7,
                                        },
                                        end: SourcePos {
                                            line: 12,
                                            col: 15,
                                        },
                                    },
                                },
                                name: Id {
                                    name: "update",
                                    pos: SourcePos {
                                        line: 12,
                                        col: 16,
                                    },
                                },
                                args: [
                                    AnnProc {
                                        proc: StringLiteral(
                                            "key1",
                                        ),
                                        span: SourceSpan {
                                            start: SourcePos {
                                                line: 12,
                                                col: 23,
                                            },
                                            end: SourcePos {
                                                line: 12,
                                                col: 29,
                                            },
                                        },
                                    },
                                    AnnProc {
                                        proc: StringLiteral(
                                            "key2",
                                        ),
                                        span: SourceSpan {
                                            start: SourcePos {
                                                line: 12,
                                                col: 31,
                                            },
                                            end: SourcePos {
                                                line: 12,
                                                col: 37,
                                            },
                                        },
                                    },
                                    AnnProc {
                                        proc: StringLiteral(
                                            "value",
                                        ),
                                        span: SourceSpan {
                                            start: SourcePos {
                                                line: 12,
                                                col: 39,
                                            },
                                            end: SourcePos {
                                                line: 12,
                                                col: 46,
                                            },
                                        },
                                    },
                                ],
                            },
                            span: SourceSpan {
                                start: SourcePos {
                                    line: 12,
                                    col: 7,
                                },
                                end: SourcePos {
                                    line: 12,
                                    col: 47,
                                },
                            },
                        },
                    },
                    span: SourceSpan {
                        start: SourcePos {
                            line: 4,
                            col: 5,
                        },
                        end: SourcePos {
                            line: 12,
                            col: 49,
                        },
                    },
                },
            },
            span: SourceSpan {
                start: SourcePos {
                    line: 3,
                    col: 1,
                },
                end: SourcePos {
                    line: 14,
                    col: 2,
                },
            },
        },
    ],
)
